File: protocol/src/main/resources/golang/ProtocolRegistrationTemplate.go
func (protocol ${protocol_name}) ProtocolId() int16 {
	return ${protocol_id}
}

func (protocol ${protocol_name}) write(buffer *ByteBuffer, packet any) {
	if packet == nil {
	    buffer.WriteInt(0)
		return
	}
	var message = packet.(*${protocol_name})
	${protocol_write_serialization}
}

func (protocol ${protocol_name}) read(buffer *ByteBuffer) any {
	var packet = new(${protocol_name})
	var length = buffer.ReadInt()
	if length == 0 {
		return packet
	}
	var beforeReadIndex = buffer.GetReadOffset()
	${protocol_read_deserialization}
	if length > 0 {
        buffer.SetReadOffset(beforeReadIndex + length)
    }
	return packet
}

File: protocol/src/main/resources/swift/ProtocolRegistrationTemplate.swift
class ${protocol_name}Registration : IProtocolRegistration {
    func write(_ buffer: ByteBuffer, _ packet: Any?) {
        if (packet == nil) {
            buffer.writeInt(0)
            return
        }
        let message = packet as! ${protocol_name}
        ${protocol_write_serialization}
    }
    
    func read(_ buffer: ByteBuffer) -> Any {
        let length = buffer.readInt()
        let packet = ${protocol_name}()
        if (length == 0) {
            return packet
        }
        let beforeReadIndex = buffer.getReadOffset()
        ${protocol_read_deserialization}
        if (length > 0) {
            buffer.setReadOffset(beforeReadIndex + length)
        }
        return packet
    }
}
